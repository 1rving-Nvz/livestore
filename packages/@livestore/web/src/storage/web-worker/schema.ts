import { Schema, Transferable } from '@livestore/utils/effect'

export class UnexpectedError extends Schema.TaggedError<UnexpectedError>()('UnexpectedError', {
  error: Schema.Any,
}) {}

export const ExecutionBacklogItemExecute = Schema.Struct({
  _tag: Schema.Literal('execute'),
  query: Schema.String,
  bindValues: Schema.Any,
})

export const ExecutionBacklogItemMutate = Schema.Struct({
  _tag: Schema.Literal('mutate'),
  mutationEventEncoded: Schema.Struct({
    mutation: Schema.String,
    args: Schema.Any,
    id: Schema.String,
  }),
})

export const ExecutionBacklogItemTxn = Schema.Struct({
  _tag: Schema.Literal('txn'),
  items: Schema.Union(ExecutionBacklogItemExecute, ExecutionBacklogItemMutate),
})

export const ExecutionBacklogItem = Schema.Union(
  ExecutionBacklogItemExecute,
  ExecutionBacklogItemMutate,
  ExecutionBacklogItemTxn,
)

export type ExecutionBacklogItem = Schema.Schema.Type<typeof ExecutionBacklogItem>

export const StorageTypeOpfs = Schema.Struct({
  type: Schema.Literal('opfs'),
  /** Default is root directory */
  directory: Schema.optional(Schema.String),
  /** Default is 'livestore-' */
  filePrefix: Schema.optional(Schema.String),
})

export const StorageTypeIndexeddb = Schema.Struct({
  type: Schema.Literal('indexeddb'),
  /** @default "livestore" */
  databaseName: Schema.optional(Schema.String),
  /** @default "livestore" */
  storeNamePrefix: Schema.optional(Schema.String),
})

export const StorageType = Schema.Union(StorageTypeOpfs, StorageTypeIndexeddb)
export type StorageType = Schema.Schema.Type<typeof StorageType>

export class InitialMessage extends Schema.TaggedRequest<InitialMessage>()(
  'InitialMessage',
  UnexpectedError,
  Schema.Void,
  { storage: StorageType },
) {}

export class ExecuteBulk extends Schema.TaggedRequest<ExecuteBulk>()('ExecuteBulk', UnexpectedError, Schema.Void, {
  items: Schema.Array(ExecutionBacklogItem),
}) {}

export class Export extends Schema.TaggedRequest<Export>()('Export', UnexpectedError, Transferable.Uint8Array, {}) {}

export class ExportMutationlog extends Schema.TaggedRequest<ExportMutationlog>()(
  'ExportMutationlog',
  UnexpectedError,
  Transferable.Uint8Array,
  {},
) {}

export class Setup extends Schema.TaggedRequest<Setup>()('Setup', UnexpectedError, Transferable.Uint8Array, {}) {}

export class Shutdown extends Schema.TaggedRequest<Shutdown>()('Shutdown', UnexpectedError, Schema.Void, {}) {}

export const Request = Schema.Union(InitialMessage, ExecuteBulk, Export, ExportMutationlog, Setup, Shutdown)
export type Request = Schema.Schema.Type<typeof Request>
