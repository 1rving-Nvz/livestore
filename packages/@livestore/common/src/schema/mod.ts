export * from './system-tables.js'
export * as DbSchema from './table-def.js'
export { SqliteAst, SqliteDsl } from './db-schema/mod.js'
export * from './mutations.js'
export * from './schema-helpers.js'
export * from './schema.js'
export * as MutationEvent from './MutationEvent.js'
export * as EventId from './EventId.js'
