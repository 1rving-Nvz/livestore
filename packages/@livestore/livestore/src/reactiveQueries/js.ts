import * as otel from '@opentelemetry/api'

import { globalDbGraph } from '../global-state.js'
import type { QueryInfo, QueryInfoNone } from '../query-info.js'
import type { Thunk } from '../reactive.js'
import type { RefreshReason } from '../store.js'
import { getDurationMsFromSpan } from '../utils/otel.js'
import type { DbContext, DbGraph, GetAtomResult, LiveQuery } from './base-class.js'
import { LiveStoreQueryBase, makeGetAtomResult } from './base-class.js'

export const computed = <TResult, TQueryInfo extends QueryInfo = QueryInfoNone>(
  fn: (get: GetAtomResult) => TResult,
  options?: {
    label: string
    dbGraph?: DbGraph
    queryInfo?: TQueryInfo
  },
): LiveQuery<TResult, TQueryInfo> =>
  new LiveStoreJSQuery<TResult, TQueryInfo>({
    fn,
    label: options?.label ?? fn.toString(),
    dbGraph: options?.dbGraph,
    queryInfo: options?.queryInfo,
  })

export class LiveStoreJSQuery<TResult, TQueryInfo extends QueryInfo = QueryInfoNone> extends LiveStoreQueryBase<
  TResult,
  TQueryInfo
> {
  _tag: 'js' = 'js'

  /** A reactive thunk representing the query results */
  results$: Thunk<TResult, DbContext, RefreshReason>

  label: string

  protected dbGraph: DbGraph

  queryInfo: TQueryInfo

  /**
   * Currently only used for "nested destruction" of piped queries
   *
   * i.e. when doing something like `const q = querySQL(...).pipe(...)`
   * we need to also destory the SQL query when the JS query `q` is destroyed
   */
  private onDestroy: (() => void) | undefined

  constructor({
    fn,
    label,
    onDestroy,
    dbGraph,
    queryInfo,
  }: {
    label: string
    fn: (get: GetAtomResult) => TResult
    /** Currently only used for "nested destruction" of piped queries */
    onDestroy?: () => void
    dbGraph?: DbGraph
    queryInfo?: TQueryInfo
  }) {
    super()

    this.onDestroy = onDestroy
    this.label = label

    this.dbGraph = dbGraph ?? globalDbGraph
    this.queryInfo = queryInfo ?? ({ _tag: 'None' } as TQueryInfo)

    const queryLabel = `${label}:results`

    this.results$ = this.dbGraph.makeThunk(
      (get, setDebugInfo, { otelTracer, rootOtelContext }, otelContext) =>
        otelTracer.startActiveSpan(`js:${label}`, {}, otelContext ?? rootOtelContext, (span) => {
          const otelContext = otel.trace.setSpan(otel.context.active(), span)
          const res = fn(makeGetAtomResult(get, otelContext))

          span.end()

          const durationMs = getDurationMsFromSpan(span)

          this.executionTimes.push(durationMs)

          setDebugInfo({ _tag: 'js', label, query: fn.toString(), durationMs })

          return res
        }),
      { label: queryLabel, meta: { liveStoreThunkType: 'jsResults' } },
    )
  }

  // pipe = <U>(fn: (result: TResult, get: GetAtomResult) => U): LiveStoreJSQuery<U> =>
  //   new LiveStoreJSQuery({
  //     fn: (get) => {
  //       const results = get(this.results$)
  //       return fn(results, get)
  //     },
  //     label: `${this.label}:js`,
  //     onDestroy: () => this.destroy(),
  //     dbGraph: this.dbGraph,
  //   })

  destroy = () => {
    this.dbGraph.destroyNode(this.results$)
    this.onDestroy?.()
  }
}
