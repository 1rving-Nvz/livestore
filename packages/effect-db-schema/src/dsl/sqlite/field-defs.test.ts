import { Schema } from '@effect/schema'
import { describe, expect, test } from 'vitest'

import * as F from './field-defs.js'

describe('FieldDefs', () => {
  test('text', () => {
    expect(F.text()).toMatchSnapshot()
    expect(F.text({})).toMatchSnapshot()
    expect(F.text({ default: null, nullable: true })).toMatchSnapshot()
    expect(F.text({ schema: Schema.literal('foo'), nullable: true, default: 'foo' })).toMatchSnapshot()
    expect(F.text({ schema: Schema.union(Schema.literal('foo')), nullable: true, default: 'foo' })).toMatchSnapshot()
  })

  test('json', () => {
    expect(F.json()).toMatchSnapshot()
    expect(F.json({ default: null, nullable: true })).toMatchSnapshot()
    expect(
      F.json({ schema: Schema.struct({ name: Schema.string }), default: { name: 'Bob' }, nullable: true }),
    ).toMatchSnapshot()
  })

  test('datetime', () => {
    expect(F.datetime()).toMatchSnapshot()
    expect(F.datetime({})).toMatchSnapshot()
    expect(F.datetime({ default: null, nullable: true })).toMatchSnapshot()
    expect(F.datetime({ default: new Date('2022-02-02') })).toMatchSnapshot()
  })

  test('boolean', () => {
    expect(F.boolean()).toMatchSnapshot()
    expect(F.boolean({})).toMatchSnapshot()
    expect(F.boolean({ default: false })).toMatchSnapshot()
  })
})
