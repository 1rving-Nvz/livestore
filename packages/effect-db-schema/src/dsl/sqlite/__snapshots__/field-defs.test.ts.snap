// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FieldDefs > boolean 1`] = `
{
  "columnType": "integer",
  "default": {
    "_id": "Option",
    "_tag": "None",
  },
  "nullable": false,
  "primaryKey": false,
  "schema": $transformOrFail {
    "ast": {
      "_tag": "Transformation",
      "annotations": {},
      "from": {
        "_tag": "NumberKeyword",
        "annotations": {
          "Symbol(@effect/schema/annotation/Description)": "a number",
          "Symbol(@effect/schema/annotation/Title)": "number",
        },
      },
      "to": {
        "_tag": "BooleanKeyword",
        "annotations": {
          "Symbol(@effect/schema/annotation/Description)": "a boolean",
          "Symbol(@effect/schema/annotation/Title)": "boolean",
        },
      },
    },
    "from": Schema {
      "ast": {
        "_tag": "NumberKeyword",
        "annotations": {
          "Symbol(@effect/schema/annotation/Description)": "a number",
          "Symbol(@effect/schema/annotation/Title)": "number",
        },
      },
      Symbol(@effect/schema/Schema): {
        "_A": [Function],
        "_I": [Function],
        "_R": [Function],
      },
    },
    "to": Schema {
      "ast": {
        "_tag": "BooleanKeyword",
        "annotations": {
          "Symbol(@effect/schema/annotation/Description)": "a boolean",
          "Symbol(@effect/schema/annotation/Title)": "boolean",
        },
      },
      Symbol(@effect/schema/Schema): {
        "_A": [Function],
        "_I": [Function],
        "_R": [Function],
      },
    },
    Symbol(@effect/schema/Schema): {
      "_A": [Function],
      "_I": [Function],
      "_R": [Function],
    },
  },
}
`;

exports[`FieldDefs > boolean 2`] = `
{
  "columnType": "integer",
  "default": {
    "_id": "Option",
    "_tag": "None",
  },
  "nullable": false,
  "primaryKey": false,
  "schema": $transformOrFail {
    "ast": {
      "_tag": "Transformation",
      "annotations": {},
      "from": {
        "_tag": "NumberKeyword",
        "annotations": {
          "Symbol(@effect/schema/annotation/Description)": "a number",
          "Symbol(@effect/schema/annotation/Title)": "number",
        },
      },
      "to": {
        "_tag": "BooleanKeyword",
        "annotations": {
          "Symbol(@effect/schema/annotation/Description)": "a boolean",
          "Symbol(@effect/schema/annotation/Title)": "boolean",
        },
      },
    },
    "from": Schema {
      "ast": {
        "_tag": "NumberKeyword",
        "annotations": {
          "Symbol(@effect/schema/annotation/Description)": "a number",
          "Symbol(@effect/schema/annotation/Title)": "number",
        },
      },
      Symbol(@effect/schema/Schema): {
        "_A": [Function],
        "_I": [Function],
        "_R": [Function],
      },
    },
    "to": Schema {
      "ast": {
        "_tag": "BooleanKeyword",
        "annotations": {
          "Symbol(@effect/schema/annotation/Description)": "a boolean",
          "Symbol(@effect/schema/annotation/Title)": "boolean",
        },
      },
      Symbol(@effect/schema/Schema): {
        "_A": [Function],
        "_I": [Function],
        "_R": [Function],
      },
    },
    Symbol(@effect/schema/Schema): {
      "_A": [Function],
      "_I": [Function],
      "_R": [Function],
    },
  },
}
`;

exports[`FieldDefs > boolean 3`] = `
{
  "columnType": "integer",
  "default": {
    "_id": "Option",
    "_tag": "Some",
    "value": false,
  },
  "nullable": false,
  "primaryKey": false,
  "schema": $transformOrFail {
    "ast": {
      "_tag": "Transformation",
      "annotations": {},
      "from": {
        "_tag": "NumberKeyword",
        "annotations": {
          "Symbol(@effect/schema/annotation/Description)": "a number",
          "Symbol(@effect/schema/annotation/Title)": "number",
        },
      },
      "to": {
        "_tag": "BooleanKeyword",
        "annotations": {
          "Symbol(@effect/schema/annotation/Description)": "a boolean",
          "Symbol(@effect/schema/annotation/Title)": "boolean",
        },
      },
    },
    "from": Schema {
      "ast": {
        "_tag": "NumberKeyword",
        "annotations": {
          "Symbol(@effect/schema/annotation/Description)": "a number",
          "Symbol(@effect/schema/annotation/Title)": "number",
        },
      },
      Symbol(@effect/schema/Schema): {
        "_A": [Function],
        "_I": [Function],
        "_R": [Function],
      },
    },
    "to": Schema {
      "ast": {
        "_tag": "BooleanKeyword",
        "annotations": {
          "Symbol(@effect/schema/annotation/Description)": "a boolean",
          "Symbol(@effect/schema/annotation/Title)": "boolean",
        },
      },
      Symbol(@effect/schema/Schema): {
        "_A": [Function],
        "_I": [Function],
        "_R": [Function],
      },
    },
    Symbol(@effect/schema/Schema): {
      "_A": [Function],
      "_I": [Function],
      "_R": [Function],
    },
  },
}
`;

exports[`FieldDefs > datetime 1`] = `
{
  "columnType": "text",
  "default": {
    "_id": "Option",
    "_tag": "None",
  },
  "nullable": false,
  "primaryKey": false,
  "schema": Schema {
    "ast": {
      "_tag": "Refinement",
      "annotations": {
        "Symbol(@effect/schema/annotation/Description)": "a valid Date",
        "Symbol(@effect/schema/annotation/Identifier)": "Date",
        "Symbol(@effect/schema/annotation/Type)": Symbol(@effect/schema/TypeId/ValidDate),
      },
      "from": {
        "_tag": "Transformation",
        "annotations": {
          "Symbol(@effect/schema/annotation/Identifier)": "DateFromString",
        },
        "from": {
          "_tag": "StringKeyword",
          "annotations": {
            "Symbol(@effect/schema/annotation/Description)": "a string",
            "Symbol(@effect/schema/annotation/Title)": "string",
          },
        },
        "to": {
          "_tag": "Declaration",
          "annotations": {
            "Symbol(@effect/schema/ArbitraryHookId)": [Function],
            "Symbol(@effect/schema/EquivalenceHookId)": [Function],
            "Symbol(@effect/schema/PrettyHookId)": [Function],
            "Symbol(@effect/schema/annotation/Description)": "a potentially invalid Date instance",
            "Symbol(@effect/schema/annotation/Identifier)": "DateFromSelf",
          },
          "typeParameters": [],
        },
      },
    },
    Symbol(@effect/schema/Schema): {
      "_A": [Function],
      "_I": [Function],
      "_R": [Function],
    },
  },
}
`;

exports[`FieldDefs > datetime 2`] = `
{
  "columnType": "text",
  "default": {
    "_id": "Option",
    "_tag": "None",
  },
  "nullable": false,
  "primaryKey": false,
  "schema": Schema {
    "ast": {
      "_tag": "Refinement",
      "annotations": {
        "Symbol(@effect/schema/annotation/Description)": "a valid Date",
        "Symbol(@effect/schema/annotation/Identifier)": "Date",
        "Symbol(@effect/schema/annotation/Type)": Symbol(@effect/schema/TypeId/ValidDate),
      },
      "from": {
        "_tag": "Transformation",
        "annotations": {
          "Symbol(@effect/schema/annotation/Identifier)": "DateFromString",
        },
        "from": {
          "_tag": "StringKeyword",
          "annotations": {
            "Symbol(@effect/schema/annotation/Description)": "a string",
            "Symbol(@effect/schema/annotation/Title)": "string",
          },
        },
        "to": {
          "_tag": "Declaration",
          "annotations": {
            "Symbol(@effect/schema/ArbitraryHookId)": [Function],
            "Symbol(@effect/schema/EquivalenceHookId)": [Function],
            "Symbol(@effect/schema/PrettyHookId)": [Function],
            "Symbol(@effect/schema/annotation/Description)": "a potentially invalid Date instance",
            "Symbol(@effect/schema/annotation/Identifier)": "DateFromSelf",
          },
          "typeParameters": [],
        },
      },
    },
    Symbol(@effect/schema/Schema): {
      "_A": [Function],
      "_I": [Function],
      "_R": [Function],
    },
  },
}
`;

exports[`FieldDefs > datetime 3`] = `
{
  "columnType": "text",
  "default": {
    "_id": "Option",
    "_tag": "Some",
    "value": null,
  },
  "nullable": true,
  "primaryKey": false,
  "schema": $union {
    "ast": {
      "_tag": "Union",
      "annotations": {},
      "types": [
        {
          "_tag": "Refinement",
          "annotations": {
            "Symbol(@effect/schema/annotation/Description)": "a valid Date",
            "Symbol(@effect/schema/annotation/Identifier)": "Date",
            "Symbol(@effect/schema/annotation/Type)": Symbol(@effect/schema/TypeId/ValidDate),
          },
          "from": {
            "_tag": "Transformation",
            "annotations": {
              "Symbol(@effect/schema/annotation/Identifier)": "DateFromString",
            },
            "from": {
              "_tag": "StringKeyword",
              "annotations": {
                "Symbol(@effect/schema/annotation/Description)": "a string",
                "Symbol(@effect/schema/annotation/Title)": "string",
              },
            },
            "to": {
              "_tag": "Declaration",
              "annotations": {
                "Symbol(@effect/schema/ArbitraryHookId)": [Function],
                "Symbol(@effect/schema/EquivalenceHookId)": [Function],
                "Symbol(@effect/schema/PrettyHookId)": [Function],
                "Symbol(@effect/schema/annotation/Description)": "a potentially invalid Date instance",
                "Symbol(@effect/schema/annotation/Identifier)": "DateFromSelf",
              },
              "typeParameters": [],
            },
          },
        },
        {
          "_tag": "Literal",
          "annotations": {
            "Symbol(@effect/schema/annotation/Identifier)": "null",
          },
          "literal": null,
        },
      ],
    },
    "members": [
      Schema {
        "ast": {
          "_tag": "Refinement",
          "annotations": {
            "Symbol(@effect/schema/annotation/Description)": "a valid Date",
            "Symbol(@effect/schema/annotation/Identifier)": "Date",
            "Symbol(@effect/schema/annotation/Type)": Symbol(@effect/schema/TypeId/ValidDate),
          },
          "from": {
            "_tag": "Transformation",
            "annotations": {
              "Symbol(@effect/schema/annotation/Identifier)": "DateFromString",
            },
            "from": {
              "_tag": "StringKeyword",
              "annotations": {
                "Symbol(@effect/schema/annotation/Description)": "a string",
                "Symbol(@effect/schema/annotation/Title)": "string",
              },
            },
            "to": {
              "_tag": "Declaration",
              "annotations": {
                "Symbol(@effect/schema/ArbitraryHookId)": [Function],
                "Symbol(@effect/schema/EquivalenceHookId)": [Function],
                "Symbol(@effect/schema/PrettyHookId)": [Function],
                "Symbol(@effect/schema/annotation/Description)": "a potentially invalid Date instance",
                "Symbol(@effect/schema/annotation/Identifier)": "DateFromSelf",
              },
              "typeParameters": [],
            },
          },
        },
        Symbol(@effect/schema/Schema): {
          "_A": [Function],
          "_I": [Function],
          "_R": [Function],
        },
      },
      Schema {
        "ast": {
          "_tag": "Literal",
          "annotations": {
            "Symbol(@effect/schema/annotation/Identifier)": "null",
          },
          "literal": null,
        },
        Symbol(@effect/schema/Schema): {
          "_A": [Function],
          "_I": [Function],
          "_R": [Function],
        },
      },
    ],
    Symbol(@effect/schema/Schema): {
      "_A": [Function],
      "_I": [Function],
      "_R": [Function],
    },
  },
}
`;

exports[`FieldDefs > datetime 4`] = `
{
  "columnType": "text",
  "default": {
    "_id": "Option",
    "_tag": "Some",
    "value": 2022-02-02T00:00:00.000Z,
  },
  "nullable": false,
  "primaryKey": false,
  "schema": Schema {
    "ast": {
      "_tag": "Refinement",
      "annotations": {
        "Symbol(@effect/schema/annotation/Description)": "a valid Date",
        "Symbol(@effect/schema/annotation/Identifier)": "Date",
        "Symbol(@effect/schema/annotation/Type)": Symbol(@effect/schema/TypeId/ValidDate),
      },
      "from": {
        "_tag": "Transformation",
        "annotations": {
          "Symbol(@effect/schema/annotation/Identifier)": "DateFromString",
        },
        "from": {
          "_tag": "StringKeyword",
          "annotations": {
            "Symbol(@effect/schema/annotation/Description)": "a string",
            "Symbol(@effect/schema/annotation/Title)": "string",
          },
        },
        "to": {
          "_tag": "Declaration",
          "annotations": {
            "Symbol(@effect/schema/ArbitraryHookId)": [Function],
            "Symbol(@effect/schema/EquivalenceHookId)": [Function],
            "Symbol(@effect/schema/PrettyHookId)": [Function],
            "Symbol(@effect/schema/annotation/Description)": "a potentially invalid Date instance",
            "Symbol(@effect/schema/annotation/Identifier)": "DateFromSelf",
          },
          "typeParameters": [],
        },
      },
    },
    Symbol(@effect/schema/Schema): {
      "_A": [Function],
      "_I": [Function],
      "_R": [Function],
    },
  },
}
`;

exports[`FieldDefs > json 1`] = `
{
  "columnType": "text",
  "default": {
    "_id": "Option",
    "_tag": "None",
  },
  "nullable": false,
  "primaryKey": false,
  "schema": Schema {
    "ast": {
      "_tag": "Transformation",
      "annotations": {},
      "from": {
        "_tag": "Transformation",
        "annotations": {},
        "from": {
          "_tag": "StringKeyword",
          "annotations": {
            "Symbol(@effect/schema/annotation/Description)": "a JSON string",
            "Symbol(@effect/schema/annotation/Identifier)": "JsonString",
            "Symbol(@effect/schema/annotation/Title)": "JsonString",
          },
        },
        "to": {
          "_tag": "UnknownKeyword",
          "annotations": {
            "Symbol(@effect/schema/annotation/Title)": "unknown",
          },
        },
      },
      "to": {
        "_tag": "AnyKeyword",
        "annotations": {
          "Symbol(@effect/schema/annotation/Title)": "any",
        },
      },
    },
    Symbol(@effect/schema/Schema): {
      "_A": [Function],
      "_I": [Function],
      "_R": [Function],
    },
  },
}
`;

exports[`FieldDefs > json 2`] = `
{
  "columnType": "text",
  "default": {
    "_id": "Option",
    "_tag": "Some",
    "value": null,
  },
  "nullable": true,
  "primaryKey": false,
  "schema": $union {
    "ast": {
      "_tag": "Union",
      "annotations": {},
      "types": [
        {
          "_tag": "Transformation",
          "annotations": {},
          "from": {
            "_tag": "Transformation",
            "annotations": {},
            "from": {
              "_tag": "StringKeyword",
              "annotations": {
                "Symbol(@effect/schema/annotation/Description)": "a JSON string",
                "Symbol(@effect/schema/annotation/Identifier)": "JsonString",
                "Symbol(@effect/schema/annotation/Title)": "JsonString",
              },
            },
            "to": {
              "_tag": "UnknownKeyword",
              "annotations": {
                "Symbol(@effect/schema/annotation/Title)": "unknown",
              },
            },
          },
          "to": {
            "_tag": "AnyKeyword",
            "annotations": {
              "Symbol(@effect/schema/annotation/Title)": "any",
            },
          },
        },
        {
          "_tag": "Literal",
          "annotations": {
            "Symbol(@effect/schema/annotation/Identifier)": "null",
          },
          "literal": null,
        },
      ],
    },
    "members": [
      Schema {
        "ast": {
          "_tag": "Transformation",
          "annotations": {},
          "from": {
            "_tag": "Transformation",
            "annotations": {},
            "from": {
              "_tag": "StringKeyword",
              "annotations": {
                "Symbol(@effect/schema/annotation/Description)": "a JSON string",
                "Symbol(@effect/schema/annotation/Identifier)": "JsonString",
                "Symbol(@effect/schema/annotation/Title)": "JsonString",
              },
            },
            "to": {
              "_tag": "UnknownKeyword",
              "annotations": {
                "Symbol(@effect/schema/annotation/Title)": "unknown",
              },
            },
          },
          "to": {
            "_tag": "AnyKeyword",
            "annotations": {
              "Symbol(@effect/schema/annotation/Title)": "any",
            },
          },
        },
        Symbol(@effect/schema/Schema): {
          "_A": [Function],
          "_I": [Function],
          "_R": [Function],
        },
      },
      Schema {
        "ast": {
          "_tag": "Literal",
          "annotations": {
            "Symbol(@effect/schema/annotation/Identifier)": "null",
          },
          "literal": null,
        },
        Symbol(@effect/schema/Schema): {
          "_A": [Function],
          "_I": [Function],
          "_R": [Function],
        },
      },
    ],
    Symbol(@effect/schema/Schema): {
      "_A": [Function],
      "_I": [Function],
      "_R": [Function],
    },
  },
}
`;

exports[`FieldDefs > json 3`] = `
{
  "columnType": "text",
  "default": {
    "_id": "Option",
    "_tag": "Some",
    "value": {
      "name": "Bob",
    },
  },
  "nullable": true,
  "primaryKey": false,
  "schema": $union {
    "ast": {
      "_tag": "Union",
      "annotations": {},
      "types": [
        {
          "_tag": "Transformation",
          "annotations": {},
          "from": {
            "_tag": "Transformation",
            "annotations": {},
            "from": {
              "_tag": "StringKeyword",
              "annotations": {
                "Symbol(@effect/schema/annotation/Description)": "a JSON string",
                "Symbol(@effect/schema/annotation/Identifier)": "JsonString",
                "Symbol(@effect/schema/annotation/Title)": "JsonString",
              },
            },
            "to": {
              "_tag": "UnknownKeyword",
              "annotations": {
                "Symbol(@effect/schema/annotation/Title)": "unknown",
              },
            },
          },
          "to": {
            "_tag": "TypeLiteral",
            "annotations": {},
            "indexSignatures": [],
            "propertySignatures": [
              {
                "annotations": {},
                "isOptional": false,
                "isReadonly": true,
                "name": "name",
                "type": {
                  "_tag": "StringKeyword",
                  "annotations": {
                    "Symbol(@effect/schema/annotation/Description)": "a string",
                    "Symbol(@effect/schema/annotation/Title)": "string",
                  },
                },
              },
            ],
          },
        },
        {
          "_tag": "Literal",
          "annotations": {
            "Symbol(@effect/schema/annotation/Identifier)": "null",
          },
          "literal": null,
        },
      ],
    },
    "members": [
      Schema {
        "ast": {
          "_tag": "Transformation",
          "annotations": {},
          "from": {
            "_tag": "Transformation",
            "annotations": {},
            "from": {
              "_tag": "StringKeyword",
              "annotations": {
                "Symbol(@effect/schema/annotation/Description)": "a JSON string",
                "Symbol(@effect/schema/annotation/Identifier)": "JsonString",
                "Symbol(@effect/schema/annotation/Title)": "JsonString",
              },
            },
            "to": {
              "_tag": "UnknownKeyword",
              "annotations": {
                "Symbol(@effect/schema/annotation/Title)": "unknown",
              },
            },
          },
          "to": {
            "_tag": "TypeLiteral",
            "annotations": {},
            "indexSignatures": [],
            "propertySignatures": [
              {
                "annotations": {},
                "isOptional": false,
                "isReadonly": true,
                "name": "name",
                "type": {
                  "_tag": "StringKeyword",
                  "annotations": {
                    "Symbol(@effect/schema/annotation/Description)": "a string",
                    "Symbol(@effect/schema/annotation/Title)": "string",
                  },
                },
              },
            ],
          },
        },
        Symbol(@effect/schema/Schema): {
          "_A": [Function],
          "_I": [Function],
          "_R": [Function],
        },
      },
      Schema {
        "ast": {
          "_tag": "Literal",
          "annotations": {
            "Symbol(@effect/schema/annotation/Identifier)": "null",
          },
          "literal": null,
        },
        Symbol(@effect/schema/Schema): {
          "_A": [Function],
          "_I": [Function],
          "_R": [Function],
        },
      },
    ],
    Symbol(@effect/schema/Schema): {
      "_A": [Function],
      "_I": [Function],
      "_R": [Function],
    },
  },
}
`;

exports[`FieldDefs > text 1`] = `
{
  "columnType": "text",
  "default": {
    "_id": "Option",
    "_tag": "None",
  },
  "nullable": false,
  "primaryKey": false,
  "schema": Schema {
    "ast": {
      "_tag": "StringKeyword",
      "annotations": {
        "Symbol(@effect/schema/annotation/Description)": "a string",
        "Symbol(@effect/schema/annotation/Title)": "string",
      },
    },
    Symbol(@effect/schema/Schema): {
      "_A": [Function],
      "_I": [Function],
      "_R": [Function],
    },
  },
}
`;

exports[`FieldDefs > text 2`] = `
{
  "columnType": "text",
  "default": {
    "_id": "Option",
    "_tag": "None",
  },
  "nullable": false,
  "primaryKey": false,
  "schema": Schema {
    "ast": {
      "_tag": "StringKeyword",
      "annotations": {
        "Symbol(@effect/schema/annotation/Description)": "a string",
        "Symbol(@effect/schema/annotation/Title)": "string",
      },
    },
    Symbol(@effect/schema/Schema): {
      "_A": [Function],
      "_I": [Function],
      "_R": [Function],
    },
  },
}
`;

exports[`FieldDefs > text 3`] = `
{
  "columnType": "text",
  "default": {
    "_id": "Option",
    "_tag": "Some",
    "value": null,
  },
  "nullable": true,
  "primaryKey": false,
  "schema": $union {
    "ast": {
      "_tag": "Union",
      "annotations": {},
      "types": [
        {
          "_tag": "StringKeyword",
          "annotations": {
            "Symbol(@effect/schema/annotation/Description)": "a string",
            "Symbol(@effect/schema/annotation/Title)": "string",
          },
        },
        {
          "_tag": "Literal",
          "annotations": {
            "Symbol(@effect/schema/annotation/Identifier)": "null",
          },
          "literal": null,
        },
      ],
    },
    "members": [
      Schema {
        "ast": {
          "_tag": "StringKeyword",
          "annotations": {
            "Symbol(@effect/schema/annotation/Description)": "a string",
            "Symbol(@effect/schema/annotation/Title)": "string",
          },
        },
        Symbol(@effect/schema/Schema): {
          "_A": [Function],
          "_I": [Function],
          "_R": [Function],
        },
      },
      Schema {
        "ast": {
          "_tag": "Literal",
          "annotations": {
            "Symbol(@effect/schema/annotation/Identifier)": "null",
          },
          "literal": null,
        },
        Symbol(@effect/schema/Schema): {
          "_A": [Function],
          "_I": [Function],
          "_R": [Function],
        },
      },
    ],
    Symbol(@effect/schema/Schema): {
      "_A": [Function],
      "_I": [Function],
      "_R": [Function],
    },
  },
}
`;

exports[`FieldDefs > text 4`] = `
{
  "columnType": "text",
  "default": {
    "_id": "Option",
    "_tag": "Some",
    "value": "foo",
  },
  "nullable": true,
  "primaryKey": false,
  "schema": $union {
    "ast": {
      "_tag": "Union",
      "annotations": {},
      "types": [
        {
          "_tag": "Literal",
          "annotations": {},
          "literal": "foo",
        },
        {
          "_tag": "Literal",
          "annotations": {
            "Symbol(@effect/schema/annotation/Identifier)": "null",
          },
          "literal": null,
        },
      ],
    },
    "members": [
      $literal {
        "ast": {
          "_tag": "Literal",
          "annotations": {},
          "literal": "foo",
        },
        "literals": [
          "foo",
        ],
        Symbol(@effect/schema/Schema): {
          "_A": [Function],
          "_I": [Function],
          "_R": [Function],
        },
      },
      Schema {
        "ast": {
          "_tag": "Literal",
          "annotations": {
            "Symbol(@effect/schema/annotation/Identifier)": "null",
          },
          "literal": null,
        },
        Symbol(@effect/schema/Schema): {
          "_A": [Function],
          "_I": [Function],
          "_R": [Function],
        },
      },
    ],
    Symbol(@effect/schema/Schema): {
      "_A": [Function],
      "_I": [Function],
      "_R": [Function],
    },
  },
}
`;

exports[`FieldDefs > text 5`] = `
{
  "columnType": "text",
  "default": {
    "_id": "Option",
    "_tag": "Some",
    "value": "foo",
  },
  "nullable": true,
  "primaryKey": false,
  "schema": $union {
    "ast": {
      "_tag": "Union",
      "annotations": {},
      "types": [
        {
          "_tag": "Literal",
          "annotations": {},
          "literal": "foo",
        },
        {
          "_tag": "Literal",
          "annotations": {
            "Symbol(@effect/schema/annotation/Identifier)": "null",
          },
          "literal": null,
        },
      ],
    },
    "members": [
      $literal {
        "ast": {
          "_tag": "Literal",
          "annotations": {},
          "literal": "foo",
        },
        "literals": [
          "foo",
        ],
        Symbol(@effect/schema/Schema): {
          "_A": [Function],
          "_I": [Function],
          "_R": [Function],
        },
      },
      Schema {
        "ast": {
          "_tag": "Literal",
          "annotations": {
            "Symbol(@effect/schema/annotation/Identifier)": "null",
          },
          "literal": null,
        },
        Symbol(@effect/schema/Schema): {
          "_A": [Function],
          "_I": [Function],
          "_R": [Function],
        },
      },
    ],
    Symbol(@effect/schema/Schema): {
      "_A": [Function],
      "_I": [Function],
      "_R": [Function],
    },
  },
}
`;
